libs/utils/energy_model.py

```diff
@@ -0,0 +1,106 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2016, ARM Limited and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import logging
+from collections import namedtuple, OrderedDict
+
+import pandas as pd
+import numpy as np
+
+from devlib.utils.misc import memoized
+
+ActiveState = namedtuple("ActiveState", ["capacity", "power"])
+ActiveState.__new__.__defaults__ = (None, None)
+
+class EnergyModelNode(namedtuple("EnergyModelNode",
+                                 ["cpus", "active_states", "idle_states",
+                                  "power_domain", "freq_domain"])):
+    @property
+    def max_capacity(self):
+        return max(s.capacity for s in self.active_states.values())
+
+    def idle_state_idx(self, state):
+        return self.idle_states.keys().index(state)
+
+EnergyModelNode.__new__.__defaults__ = (None, None, None, None, None)
+
+class PowerDomain(object):
+    def __init__(self, idle_states, parent, cpus):
+        self.cpus = set()
+        self.idle_states = idle_states
+
+        self.parent = parent
+        self.add_cpus(set(cpus))
+
+    def add_cpus(self, cpus):
+        self.cpus = self.cpus.union(cpus)
+        if self.parent:
+            self.parent.add_cpus(self.cpus)
+
+    def __repr__(self):
```

Comment:
EnergyModelNode inherits a `__repr__` fom `namedtuple`. Good point on EnergyModel - I do have one in a subsequent patch but I'll need to think about how it should look, it's a little bit awkward because it hold quite a lot of data so the same style as the `namedtuple` `__repr__` might be totally unreadable.